sub: add SDH subtitle filter

DanOscarsson opened the pull request:
Add subtitle filter to remove additions for deaf or hard-of-hearing
(SDH). This is for English, but may in part work for others too.
This is an ASS filter and the intention is that it can always be
enabled as it by default do not remove parts that maybe normal text.
Harder filtering can be enabled with an additional option.

I agree that my changes can be relicensed to LGPL 2.1 or later.

wm4 commented:
A rather specific feature full of heuristics, complicated C-style string processing that you need to look extremely carefully at to exclude security-relevant bugs (and will probably have some anyway), duplicated code paths - unfortunately I'm not very happy with this PR.

DanOscarsson commented:
This type of filtering is based on heuristics. No way to avoid that.
Yes, I agree that there is complicated C string handling and that you can easily get it wrong. Took a little while before it worked (have used for daily for about a month). But yes, there may be some bugs left.
It can probably be written in a more secure way and broken into parts, though it will probably be a bit slower and need temporary memory, And one could use a regular expression library to do substitutions with. Unfortunately I have too many video with embedded SDH so I need something like this.
If you prefer we can close this one, but as it works fine for me I have other things I want to fix before looking at some way to make this better.

wm4 commented:
It can probably go in, but will need some cleanup.

DanOscarsson commented:
OK. What cleanup do you want?
When I added the option to filer harder some code is more or less duplicated as I did not want to complicate the tests for allowed characters any more.
If I have some examples of cleanup you prefer, I can try to cleanup the code - hopefully I will not introduce any new bugs.

wm4 commented on sub/sd.h:
```diff
@@ -45,6 +45,7 @@ struct sd_functions {
 struct lavc_conv;
 struct lavc_conv *lavc_conv_create(struct mp_log *log, const char *codec_name,
                                    char *extradata, int extradata_len);
+void filter_SDH(struct sd *sd, char * format, int n_ignored, char * ass, int filter_harder);
```
Filters should pick up filter-specific options on their own, instead of requiring passing them to it every time.

wm4 commented on sub/sd_ass.c:
```diff
@@ -272,9 +278,30 @@ static void decode(struct sd *sd, struct demux_packet *packet)
+            ass_process_chunk(track, packet->buffer, packet->len,
+                              llrint(packet->pts * 1000),
+                              llrint(packet->duration * 1000));
+        }
```
You duplicate the ass_process_chunk 3 times -> unacceptable.

wm4 replied:
I still think this should be solved.

wm4 commented on sub/sd_ass.c:
```diff
@@ -272,9 +278,30 @@ static void decode(struct sd *sd, struct demux_packet *packet)
+                                  llrint(packet->duration * 1000));
+            } else {
+                memcpy(ctx->fbuffer, packet->buffer, packet->len);
+                ctx->fbuffer[packet->len] = '\0';
```
I'm not sure what this buffer management is about, but it probably shouldn't exist.

DanOscarsson replied:
My filter need a null terminated string and I wanted to avoid allocating memory, calling filter and then discarding memory for every subtitle. So the buffer was for having data that could be reused for all subtitle lines. But I think that in mpv it is very common to allocate mamory, copy data there and the discard it again instead of trying to avoid copying/allocating.

wm4 replied:
Yes, it's not going to matter for low-bitrate subtitles. libass does a lot of allocations too.

wm4 commented on sub/sd_ass.c:
```diff
@@ -272,9 +278,30 @@ static void decode(struct sd *sd, struct demux_packet *packet)
+            if (packet->len + 1 > ctx->fbuffer_len) {
+                free(ctx->fbuffer);
+                ctx->fbuffer_len = (((packet->len + 1) >> 8) + 1) << 8;
+                ctx->fbuffer = malloc(ctx->fbuffer_len);
```
We have tons of helpers to avoid such code.

wm4 commented on sub/sd_ass.c:
```diff
@@ -259,8 +261,12 @@ static void decode(struct sd *sd, struct demux_packet *packet)
-            ass_process_data(track, r[n], strlen(r[n]));
+        for (int n = 0; r && r[n]; n++) {
+            if (sd->opts->sub_filter_SDH)
+                filter_SDH(sd, track->event_format, 0, r[n] , sd->opts->sub_filter_SDH_harder);
```
This permutes the buffer allocated/provided by libavcodec. I'm not sure if that's allowed, but it sure isn't nice.

DanOscarsson replied:
Maybe not, but it avoids allocating memory just for a possibly rewritten subtitle line.
One way could be to have the filter only read original line and allocate a new and write to it.
Combining that with the call for ASS subtitles to the filter where you may prefer that I do not use a reusable buffer, there would be one more allocate/copy that is immediately discarded.
To avoid that I would have to change the filter to instead of having null terminated input string use length to tell when subtitle ends.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,240 @@
+// Future enhancement could be to add "- " when a speaker label
+// is removed. Need to check what is used in current file as some
+// use "-" and some nothing at all.
+void filter_SDH(struct sd *sd, char * format, int n_ignored, char * ass, int filter_harder)
```
This doesn't explain what the parameters mean, or that the function writes to some.

wm4 replied:
Also `type * name` is against the coding style.

DanOscarsson replied:
It does say that the incoming line in "ass" is overwritten in place. But yes, it could be done better.
I looked around at other parts of mpv and there were little comments on functions anywhere and even less places where I could see how you prefer to explain parameters. So I did not explain them here.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,240 @@
+        // skip leading ass formatting like bold or italic
+        while (*p == '{') {
+            *t++ = *p++;
+            for (; *p && *p != '}'; *t++ = *p++);
```
This is a good example of C string code that's both dumb and too clever at the same time.
- it has a redundant check (`*p` is either 0 or '}', why check for both?)
- it doesn't use the first expression of the `for`
- it mutates the data in the third expression of the `for`
- it has no `for` body (good example for how to use `for` how to obfuscate)
- there's nothing that guarantees that `t` doesn't go past its allocation, or that `t` doesn't stomp over `p` before it's read

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,240 @@
+            *t++ = *p++;
+            for (; *p && *p != '}'; *t++ = *p++);
+            if (*p)
+                *t++ = *p++;
```
Why does it even do that? Why copy one character and not two?

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,240 @@
+            for (c = p; *c && *c != ':' && ((*c >= 'A' && *c <= 'Z') || (*c >= 'a' && *c <= 'z') ||
+                                            (*c >= '0' && *c <= '9') ||
+                                            *c == ' ' || *c == '\'' ||
+                                            *c == '#' || *c == '.' || *c == ',') ; c++);
```
And here's where I stop reading because it becomes too awful. At least it could have used ctypes.

DanOscarsson replied:
I have now done some timing tests (which I should have done before) and ctype functions are as fast so it will not slow it down. I will change to ctype functions.

wm4 commented:
The C string processing in your filter function is just awful. I'm not sure how to improve this. I found that the bstr functions usually help with string processing, but I don't know if the result would be better when using that.

DanOscarsson commented:
Some cleanup is done, more comments so the code may be easier to understand.
If you do not like my for statements with no body I can rewrite them into while statements.
Hope it looks better now.

wm4 commented on sub/filter.c:
```diff
@@ -28,119 +30,129 @@
-            *t++ = *p++;
-            for (; *p && *p != '}'; *t++ = *p++);
+            // copy "{text", stop on "}" or string end if string ends before "}" is found
+            for (*t++ = *p++; *p && *p != '}'; *t++ = *p++);
```
I can't accept code like this.

DanOscarsson commented:
Changed the for loop and several others.
If it does still not look they way you prefer, please give me an example on how you want it to be written.

wm4 commented:
> please give me an example on how you want it to be written.

It got to stop being a deeply nested 200 lines function with intricate string processing. I'm not sure what else to say. Maybe you could abstract away some thing with functions, and maybe not writing to the same string you're parsing helps. I don't know.

DanOscarsson commented:
I have now done a major rewrite of the filter.
So far it work well, but my previous version was better tested as I had used it for a month.
Woeks for those videos I have tested and for the normal viewing the last 2 days.
If the code looks better for you now, do you want to add it now or shall we close this request and I can submit it again after having used it myself some more weeks?

wm4 commented:
Better, but still not sufficient.
- you write to a destination buffer, without bounds checking
- there is apparently duplicated code wrt. detecting ASS override tags (with sd_ass.c)
- undefined behavior like `char c = ...; isdigit(c);` (these functions take an int in range 0-255, while char can be signed and in the range -128-127)
- probably more

DanOscarsson commented:
- No need to check bounds for destination buffer. The destination buffer has the same size as source plus one and as to it is copied all non skipped characters from source, never more in then in source. So why is bounds checking needed? The only extra character is terminating \0 which there is space for in the buffer. Or you want me to have destination buffer like the one you use in sd_ass for ass_to_plaintext?
- Duplicated code - yes I see that sd_ass do some checking for if override tags exist but my code copies or skips the tags so I do not think I can use that code.
- The char is now unsigned so isdigit and others should work.

wm4 commented:
> No need to check bounds for destination buffer. The destination buffer has the same size as source plus one and as to it is copied all non skipped characters from source, never more in then in source. So why is bounds checking needed? The only extra character is terminating \0 which there is space for in the buffer. Or you want me to have destination buffer like the one you use in sd_ass for ass_to_plaintext?

Checking it is better, because it's more robust.

> Duplicated code - yes I see that sd_ass do some checking for if override tags exist but my code copies or skips the tags so I do not think I can use that code.

That probably means you handle some obscure cases not correctly, but I didn't check too closely.

> The char is now unsigned so isdigit and others should work.

I just checked how other mpv code handles it, and it turns out we have `misc/ctype.h`, and all code uses it consistently (except one case, which I just fixed). Might be better to use that, and then you don't need unsigned char everywhere.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,420 @@
+    // skip trailing spaces
+    while (rp < end && *rp == ' ') {
+        rp++;
+    }
```
Maybe strspn?

DanOscarsson replied:
Not a null terminated string. Also should normally only be one space so the overhead would be large.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,420 @@
+        skip_speaker_label(sd, &rp, end, &wp);
+
+        // go through the rest of the line looking for SDH in () or []
+        while (rp < end && !(rp < end-1 && *rp == '\\' && rp[1] == 'N')) {
```
Is it possible to write `rp[0]` consistently instead of `*rp`?

DanOscarsson replied:
Yes, I can do that. Though I see that other files in mpv does not do that. For example in osd_libass.c

wm4 commented:
Is it possible to avoid pointer comparisons for strings?

DanOscarsson commented:
Yes I see that you have your own ctype. It implements the same tests that I replaced when you wanted me to use ctype instead. I can change to your implementation though my tests showed that the standard ctype is faster then implementing them through comparison tests which your code does.

The override tag code you have in sd_ass does not look like it works the same way I need. So I cannot see any easy way to use it.

I can add bound checking on write buffer too, but it will slow down the code and probably should have error checking and notification. I will have to look at it. But the intention was that it is not needed as filtered text cannot be longer then original text and write buffer is same length as original text plus one.

By pointer comparison you mean my tests like: rp < end?
When I did the rewrite I changed the code so it handles text that is limited by length instead of null (as one of the calls from sd_ass has it that way) so now instead of testing for *rp I need to test for end of string by rp < end.

DanOscarsson commented:
Now the SDH filter is more like the code in sd_ass with a null terminated input string and bounds checking when writing to output buffer.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,468 @@
+        return;
+    }
+    buf->pos = pos;
+}
```
What's the purpose of set_pos/get_pos?

DanOscarsson replied:
set_pos is to change write position in buffer. Needed when some written data have to be removed as it was detected after filtering that it was not needed. set_pos checks so that position is not set out of bounds. get_pos was just to have a get when there was a set.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,468 @@
+        buf->string[buf->pos++] = c;
+    } else {
+        MP_WARN(sd, "Internal filter error - append bad pos = %d, length = %d\n",
+                buf->pos, buf->length);
```
It would be better to just set an error flag or so. Or maybe ignore unlikely errors entirely (as long as they don't trigger UB).

DanOscarsson replied:
I have changed to MP_DBG for the internal errors and MP_VERBOSE for the cases where filtering is not possible due to missing/bad input data. But I can remove them totally if that is preferred.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,468 @@
+    } else {
+        // non space follows - no speaker label
+        set_pos(sd, buf, old_pos);
+        return;
```
Is it intended that *rpp is not set before return?

DanOscarsson replied:
Yes. As is said in comments for function read pointer and write position is not changed if the function could not remove any data.

wm4 commented on sub/filter.c:
```diff
@@ -0,0 +1,468 @@
+                   rp[0] == '-' || rp[0] == '"' || rp[0] == '\\') {
+            if (!mp_isdigit(rp[0])) {
+                only_digits = false;
+            }
```
Redundant `{` / `}`, see coding style rules.

DanOscarsson replied:
Ok. Removed.

wm4 commented:
I'd say it's heading into the right direction. With some more work this can be merged.

In addition, there's some debugging code and non-English comments in there.

DanOscarsson commented:
Sorry. Missed to remove my debugging code. Fixed.
I also got a comment on sd_ass through e-mail but that was for an old version of the code.

DanOscarsson commented:
Added so that terminating \0 gets written even when buffer is filled.
I see that it is handled in sd_ass.c too - in a slightly different way.

DanOscarsson commented:
I have been using this code daily with no problems. Is there anything more that needs to be done before it can be merged?

wm4 commented:
Issues I still see:
- the ass_process* calls are duplicated (so now there are 4 places where we pass the subtitle to libass, where there should be only 1 (the current minimum is 2 due to ffmpeg API weaknesses)
- the "internal error" debug mp_msgs should probably be removed
- get_pos function still exists

DanOscarsson commented:
Removed the "internal error" messages and the get_post function.
Now ass_process* is only called in 2 places. Hope the way I did it looks ok.

wm4 commented on sub/sd_ass.c:
```diff
@@ -281,21 +279,20 @@ static void decode(struct sd *sd, struct demux_packet *packet)
         if (sd->opts->sub_filter_SDH) {
-            char *indata = talloc_strndup(NULL, packet->buffer, packet->len);
-            char *ass_line = filter_SDH(sd, track->event_format, 1, indata);
+            char *indata = talloc_strndup(NULL, ass_line, ass_len);
```
Why does it do this?

DanOscarsson replied:
I assumed that in this case the data in packet->buffer is not \0 terminated.
And as the filter expects a \0 terminated string I have to duplicate the data and \0 terminate it.

wm4 replied:
Then the filter_sdh function could do this strdup internally.

DanOscarsson replied:
So you prefer that I add a parameter (a length) to the filter that signals if the input string is null terminated or not? The lavc_conv_decode returns a \0 terminated string so in that case it can be used directly.

wm4 replied:
Yes, you could pass strlen()+1 in the lavc_conv_decode case, and the filter code could check the last byte to see if it's 0 or not.

DanOscarsson commented:
Now the filter handles both null terminated input and by length specified.

wm4 commented:
Can you squash and rebase this? Then I can merge.

wiiaboo commented on DOCS/man/options.rst:
```diff
@@ -1991,6 +1991,23 @@ Subtitles
+
+``--sub-filter-sdh-harder=<yes|no>``
+    Do harder SDH filtering (if enabled by ``--sub-filter-sdh``).
+    Will also remove speker labels and text within parentheses using both
```
speaker*

DanOscarsson replied:
Yes you are right.
Fixed. Squashed again.

wm4 commented:
Merged. Renamed filter.c to filter_sdh.c, and moved a declaration, no other changes by me.
